functions/src/functions/users.ts

```diff
@@ -0,0 +1,340 @@
+            email: user.email,
+            phoneNumber: user.phoneNumber,
+            photoURL: user.photoURL,
+          },
+        }
+        if (request.data.includeClinicianData ?? false) {
+          const clinician = await firestoreService.getClinician(userId)
+          userInformation.clinician = clinician.content
+        }
+        if (request.data.includePatientData ?? false) {
+          const patient = await firestoreService.getPatient(userId)
+          userInformation.patient = patient.content
+        }
+        if (request.data.includeUserData ?? false) {
+          userInformation.user = userData.content
+        }
+        result[userId] = { data: userInformation }
+      } catch (error) {
+        if (error instanceof https.HttpsError) {
+          result[userId] = {
+            error: {
+              code: error.code,
+              message: error.message,
+            },
+          }
+        } else if (error instanceof Error) {
+          result[userId] = {
+            error: {
+              code: '500',
+              message: error.message,
+            },
+          }
+        } else {
+          result[userId] = {
+            error: {
+              code: '500',
+              message: 'Internal server error',
+            },
+          }
+        }
+      }
+    }
+    return result
+  },
+)
+
+export interface UpdateUserInformationInput {
+  userId?: string
+  data?: UserInformation
+}
+
+export const updateUserInformationFunction = onCall(
+  async (request: CallableRequest<UpdateUserInformationInput>) => {
+    if (!request.auth?.uid)
+      throw new https.HttpsError('unauthenticated', 'User is not authenticated')
+
+    if (!request.data.userId)
+      throw new https.HttpsError('invalid-argument', 'User ID is required')
+
+    if (!request.data.data)
+      throw new https.HttpsError('invalid-argument', 'User data is required')
+
+    const firestoreService = new FirestoreService()
+    const user = await firestoreService.getUser(request.data.userId)
+    const organization = user.content?.organization
+
+    const securityService = new SecurityService()
+    try {
+      await securityService.ensureClinician(request.auth, organization)
+    } catch {
+      securityService.ensureUser(request.auth, request.data.userId)
+    }
+
+    const auth = admin.auth()
+    await auth.updateUser(request.data.userId, {
+      displayName: request.data.data.auth.displayName,
+      email: request.data.data.auth.email,
+      phoneNumber: request.data.data.auth.phoneNumber,
+      photoURL: request.data.data.auth.photoURL,
+    })
+  },
+)
+
+export interface DeleteUserInput {
+  userId?: string
+}
+
+export const deleteUserFunction = onCall(
+  async (request: CallableRequest<DeleteUserInput>) => {
+    if (!request.auth?.uid)
+      throw new https.HttpsError('unauthenticated', 'User is not authenticated')
+
+    if (!request.data.userId)
+      throw new https.HttpsError('invalid-argument', 'User ID is required')
```

Comment:
For validating input on the server, I like to use [zod](https://github.com/colinhacks/zod). With little work to write a wrapper that results with Firebase-compliant errors (`validate` in the example), imaginary API here could look like this:


```
const deleteUserInputSchema = z.object({
  userId: z.string()
})

// type inferred directly from validation schema, ensuring 100% run-time safety and types in-sync with validation
type DeleteUserInput = z.Infer<typeof deleteUserInputSchema> 

export const deleteUserFunction = onCall(async (request) => {
  const data = validate(deleteUserInputSchema, request.data)
  // data is 100% safe to be DeleteUserInput type
})
```

Reply:
Nice - thank you for the suggestion! I think I will integrate it in a future PR, hope that's okay

Reply:
Sounds like a good future addition 👍